release/pkg/manager/calico/manager.go

```diff
@@ -462,22 +511,199 @@ func (r *CalicoManager) releasePrereqs() error {
+	}
+
+	// If we are releasing to projectcalico/calico, make sure we are releasing to the default registries.
+	if r.githubOrg == "projectcalico" && r.repo == "calico" {
+		if !reflect.DeepEqual(r.imageRegistries, defaultRegistries) {
+			return fmt.Errorf("image registries cannot be different from default registries for a release")
+		}
+	}
+
+	return r.assertImageVersions()
+}
+
+type imageExistsResult struct {
+	name   string
+	image  string
+	exists bool
+	err    error
+}
+
+func imgExists(name string, component registry.Component, ch chan imageExistsResult) {
+	r := imageExistsResult{
+		name:  name,
+		image: component.String(),
+	}
+	r.exists, r.err = registry.ImageExists(component.ImageRef())
+	ch <- r
+}
+
+// Check that the environment has the necessary prereqs for publishing hashrelease
+func (r *CalicoManager) hashreleasePrereqs() error {
+	if r.publishHashrelease {
+		if !r.hashreleaseConfig.Valid() {
+			return fmt.Errorf("missing hashrelease server configuration")
+		}
+	}
+	images, err := pinnedversion.RetrieveComponentsToValidate(r.tmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to get components to validate: %s", err)
+	}
+	if r.publishImages {
+		return r.assertImageVersions()
+	} else {
+		results := make(map[string]imageExistsResult, len(images))
+		ch := make(chan imageExistsResult)
+		for name, component := range images {
+			go imgExists(name, component, ch)
+		}
+		for range images {
+			res := <-ch
+			results[res.name] = res
+		}
+		failedImageList := []string{}
+		for name, r := range results {
+			logrus.WithFields(logrus.Fields{
+				"image":  r.image,
+				"exists": r.exists,
+			}).Info("Validating image")
+			if r.err != nil || !r.exists {
+				logrus.WithError(r.err).WithField("image", name).Error("Error checking image")
+				failedImageList = append(failedImageList, images[name].String())
+			} else {
+				logrus.WithField("image", name).Info("Image exists")
+			}
+		}
+		failedCount := len(failedImageList)
+		if failedCount > 0 {
+			return fmt.Errorf("failed to validate %d images: %s", failedCount, strings.Join(failedImageList, ", "))
+		}
+	}
+	if r.imageScanning {
+		logrus.Info("Sending images to ISS")
+		imageList := []string{}
+		for _, component := range images {
+			imageList = append(imageList, component.String())
+		}
+		imageScanner := imagescanner.New(r.imageScanningConfig)
+		err := imageScanner.Scan(imageList, r.hashrelease.Stream, false, r.tmpDir)
+		if err != nil {
+			// Error is logged and ignored as this is not considered a fatal error
+			logrus.WithError(err).Error("Failed to scan images")
+		}
 	}
+	return nil
+}
 
-	// TODO: Make sure the environment isn't dirty.
+// Check that the images exists with the correct version.
+func (r *CalicoManager) assertImageVersions() error {
+	for _, img := range images {
+		imageName := strings.TrimPrefix(img, "calico/")
+		switch img {
+		case "calico/apiserver":
+			for _, reg := range r.imageRegistries {
```

Comment:
I think the main reasons for the switch statement is because the command to check the version is different, right?

Could we factor out all of the common code here?

```go
for _, img := range images {
  switch img {
    case "calico/apiserver":
      cmd = []string{...}
    case "calico/cni": 
      cmd = []string{...}
  }
  for _, reg := r.imageRegistries {
    // Run command determined above. 
  }
}
```

Reply:
Not really, it may be more cumbersome and confusing for a few reasons:

- apiserver: always return an error for the version check as there is no kube server
- each command is formatted based on the registry so i will have to return a cmdFornat instead of command.
- cni: we check the version of both calico and calico-ipam which will require returning an array of command
- key-cert-provisioner images (key-cert-provisioner & test-signer) do not have versions in them atm

Having it separated as-is is more readable IMO instead